netCDF4/_netCDF4.pyx

```diff
@@ -2273,14 +2289,36 @@ version 4.1.2 or higher of the netcdf C lib, and rebuild netcdf4-python."""
         # view.obj is checked, ref on obj is decremented and obj will be null'd out
         PyBuffer_Release(&self._buffer)
 
+    def _close_mem(self, check_err):
+        cdef NC_memio memio
+        cdef int ierr
+
+        ierr = nc_close_memio(self._grpid, &memio)
+
+        if check_err:
+            _ensure_nc_success(ierr)
+
+        self._isopen = 0 # indicates file already closed, checked by __dealloc__
+
+        # Only release buffer if close succeeded
+        # per impl of PyBuffer_Release: https://github.com/python/cpython/blob/master/Objects/abstract.c#L667
+        # view.obj is checked, ref on obj is decremented and obj will be null'd out
+        PyBuffer_Release(&self._buffer)
+
+        # return bytes representing in-memory dataset
+        return PyBytes_FromStringAndSize(<char *>memio.memory, memio.size)
```

Comment:
Two downsides of this approach:
1. You have to close a dataset to copy it's memory contents -- but you might like to copy the data without closing the dataset just yet.
2. This will always make a copy of the Dataset's memory into a Python `bytes` object. There's no way to read data without copying all of it.

Instead, what about exposing the contents of `self._buffer` via Python's Buffer interface? My inclination would be to write a `netCDF4.Dataset.buffer` property that returns an object satisfying CPython's buffer protocol. I think this would be [pretty easy in Cython](https://cython.readthedocs.io/en/latest/src/userguide/buffer.html) and would no-copy access in Python, e.g.,by creating a `BytesIO` object like `io.BytesIO(dataset.buffer)`.

Reply:
You have a point - will think about this some more.

Reply:
The other reason why I like the attribute is that you can still use contextmanger for the file, e.g.,
```python
with netCDF4.Dataset('somewhere', mode='w', diskless=True) as ds:
    ...
bytes(ds.buffer)  # could also make this a method, e.g., get_buffer() or get_file_image() (like pytables)
```

One minor correction: clean-up should happen in the `__dealloc__` or `__del__` method of the object holding the buffer, in case it has a longer lifetime than the Dataset.

It looks like this StackOverflow post has a good description of how to make these objects:
https://stackoverflow.com/a/28166272/809705

Reply:
Working on a solution that returns a memoryview object instead of a bytes object (without making a copy).  See https://stackoverflow.com/questions/28160359/how-to-wrap-a-c-pointer-and-length-in-a-new-style-buffer-object-in-cython

Don't really see the advantage of keeping the memoryview as an attribute of the Dataset object - seems counterintutive to keep using the Dataset after it's closed.

Reply:
Sorry I didn't fully read your comment above - you already referenced that stackoverflow and pointed out that keeping the buffer as an attribute lets you use the context manager.